content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc

```diff
@@ -0,0 +1,399 @@
+        "dev.morling.links.base.LinksMessages", context.getLocale());
+
+    return "[User: " + context.getName() + "] " + bundle.getString(key);
+  }
+}
+----
+
+Here, the error handler simply obtains the message for a given key from the bundle,
+using the locale of some user context object, and returning a message prefixed with the user's name.
+This implementation just serves for example purposes of course;
+in an actual application, message keys might for instance be obtained from application specific exception types,
+raised in the different modules, and logged in a unified way via the error handler.
+
+== Resource Bundle Providers
+
+With the code in the _core_ module in place
+(mostly, that is, as we'll see in a bit),
+let's shift our attention towards the resource bundle providers in the different application modules.
+Not too suprising, they need to define an implementation of the `LinksMessagesProvider` contract.
+
+There is one challenge though:
+how can the different modules contribute implementations for one and the same bundle base name and locale?
+Once the look-up code in `ResourceBundle` has found a provider which returns a bundle for a requested name and locale,
+it will not query any other bundle providers.
+In our case though, we need to be able to obtain messages from any of the bundles contributed by the different modules:
+messages related to green keeping must be obtained from the bundle of the `dev.morling.links.greenkeeping` module,
+tournament messages from `dev.morling.links.tournament`, and so on.
+
+The idea to address this concern is the following:
+
+* Prefix each message key with a module specific string, resulting in keys like `tournament.fullybooked`, `greenkeeping.greenclosed`, etc.
+* When requesting the bundle for a given key in the error handler class, obtain the key's prefix and pass it to bundle providers
+* Let bundle providers react only to their specific message prefix
+
+This is where things become a little bit fiddly:
+there isn't a really good way for passing such contextual information from bundle consumers to providers.
+Our loop hole here will be to squeeze that information into the the requested https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Locale.html[`Locale`] instance.
+Besides the well-known language and country attributes, `Locale` can also carry _variant_ data and even application specific https://docs.oracle.com/javase/tutorial/i18n/locale/extensions.html[_extensions_].
+
+The latter would actually be pretty much ideal for our purposes.
+But unfortunately, extensions aren't evaluated by the look-up routine in `ResourceBundle`.
+So instead we'll go with propagating the key namespace information via the locale's variant.
+First, let's revisit the code in the `ErrorHandler` class:
+
+[source,java,linenums=true]
+----
+public class ErrorHandler {
+
+  public String getErrorMessage(String key, UserContext context) {
+      String component = key.split("\\.")[0]; // <1>
```

Comment:
nit pick -> why is the var named 'component'. if I didn't miss it the text never talks about components but bundles, modules and key prefixes?

Reply:
Good catch, I was/am using the terms component and modules somewhat interchangeable. I'll adjust it.